simulator/ssh/client.go

```diff
@@ -0,0 +1,211 @@
+	}
+	r, err := sess.StdoutPipe()
+	if err != nil {
+		sess.Close()
+		sshClient.Close()
+		return nil, fmt.Errorf("failed initializing SFTP IO: %v", err)
+	}
+	return &Client{sshClient, sess, w, r}, nil
+}
+
+// Teardown closes the underlying SSH client and session.
+func (c *Client) Teardown() {
+	// ssh.Client.Close should close underlying net.Conn.
+	c.sshClient.Close()
+	// ssh.Session.Close will close channel used as the read/write pipes.
+	c.sess.Close()
+}
+
+// ReadResp reads server responses to client requests.  A structure satisfying the Packet
+// interface is returned along with an error.  The returned structure will either match
+// the expectedRespType, or will be nil with an error.  If expectedRespType is a status,
+// the caller is expected to check if the status is carrying an error.
+func (c *Client) ReadResp(expectedRespType uint8) (sdipacket.Packet, error) {
+	resp, err := ReadPacket(c.r)
+	if err != nil {
+		return nil, err
+	}
+	// First byte carries the packet type code, bytes 1 onward carry the payload.
+	respType := resp[0]
+	respData := resp[1:]
+	// Handle status packets first, regardless whether that's the expected response, as
+	// they may signal an error.
+	if respType == sdifxp.TypeCodeStatus {
+		// Unmarshalling may fail.
+		statusResp, err := StatusResp(respData)
+		if err != nil {
+			return nil, err
+		}
+		// If the caller was expecting a status response, do no further processing.  The
+		// caller is responsible for determining if this is an error.
+		if expectedRespType == respType {
+			return statusResp, nil
+		}
+		// Otherwise, check to see if the status response carries an error, and return it
+		// along with a nil sdipacket.Packet.  We could add some additional error prefix
+		// to the message, but it's becoming crowded with little added informational value.
+		if statusResp.ErrCode != SSH_FX_OK {
+			return nil, fmt.Errorf("%v", statusResp.ErrMsg)
+		}
+		// Otherwise, this appears to be an invalid response.
+		return nil, fmt.Errorf("unexpected response type")
+	}
+	// We received something other than a status response.  It better be what the caller
+	// wants.
+	if expectedRespType != respType {
+		return nil, fmt.Errorf("unexpected response type")
+	}
+	switch respType {
+	case sdifxp.TypeCodeVersion:
+		return VersionResp(respData)
+	case sdifxp.TypeCodeHandle:
+		return OpenResp(respData)
+	default:
+		return nil, fmt.Errorf("unsupported response type")
+	}
+}
+
+// SendInit sends an init request to the server, returning a Version and an error.
+func (c *Client) SendInit() (*sdifxp.Version, error) {
+	initPkt := sdifxp.Init{Version: ClientVer}
+	rawInitPkt := sdipacket.MakeRawPacket(&initPkt)
+	c.w.Write(rawInitPkt.Marshal())
+	resp, err := c.ReadResp(sdifxp.TypeCodeVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed init: %v", err)
+	}
+	if versionResp, ok := resp.(*sdifxp.Version); ok {
+		return versionResp, nil
+	}
+	panic(fmt.Errorf("failed init: invalid response processed"))
```

Comment:
why panic instead of an error? We don't want to crash flightsim in case of misbehaving ssh servers.

Reply:
:+1: